services/src/datasets/external/sentinel_s2_l2a_cogs.rs

```diff
@@ -0,0 +1,640 @@
+    api_url: String,
+    #[allow(dead_code)]
+    meta_data: SentinelMetaData,
+    datasets: HashMap<DatasetId, SentinelDataset>,
+}
+
+impl SentinelS2L2aCogsDataProvider {
+    pub fn new(id: DatasetProviderId, api_url: String) -> Self {
+        let meta_data = Self::load_metadata();
+        Self {
+            id,
+            api_url,
+            datasets: Self::create_datasets(&id, &meta_data),
+            meta_data,
+        }
+    }
+
+    fn load_metadata() -> SentinelMetaData {
+        // TODO: fetch dataset metadata from config or remote
+        // TODO: is there a no data value?
+        SentinelMetaData {
+            bands: vec![
+                Band::new("B01".to_owned(), Some(0.), RasterDataType::U16),
+                Band::new("B02".to_owned(), Some(0.), RasterDataType::U16),
+            ],
+            zones: vec![
+                Zone::new("UTM32N".to_owned(), 32632),
+                Zone::new("UTM36S".to_owned(), 32736),
+            ],
+        }
+    }
+
+    fn create_datasets(
+        id: &DatasetProviderId,
+        meta_data: &SentinelMetaData,
+    ) -> HashMap<DatasetId, SentinelDataset> {
+        meta_data
+            .zones
+            .iter()
+            .flat_map(|zone| {
+                meta_data.bands.iter().map(move |band| {
+                    let dataset_id: DatasetId = ExternalDatasetId {
+                        provider: *id,
+                        id: format!("{}:{}", zone.name, band.name),
+                    }
+                    .into();
+                    let listing = DatasetListing {
+                        id: dataset_id.clone(),
+                        name: format!("Sentinel S2 L2A COGS {}:{}", zone.name, band.name),
+                        description: "".to_owned(),
+                        tags: vec![],
+                        source_operator: "GdalSource".to_owned(),
+                        result_descriptor: RasterResultDescriptor {
+                            data_type: band.data_type,
+                            spatial_reference: SpatialReference::new(
+                                SpatialReferenceAuthority::Epsg,
+                                zone.epsg,
+                            )
+                            .into(),
+                            measurement: Measurement::Unitless,
+                            no_data_value: band.no_data_value,
+                        }
+                        .into(),
+                    };
+
+                    let dataset = SentinelDataset {
+                        zone: zone.clone(),
+                        band: band.clone(),
+                        listing,
+                    };
+
+                    (dataset_id, dataset)
+                })
+            })
+            .collect()
+    }
+}
+
+#[async_trait]
+impl DatasetProvider for SentinelS2L2aCogsDataProvider {
+    async fn list(&self, _options: Validated<DatasetListOptions>) -> Result<Vec<DatasetListing>> {
+        // TODO: options
```

Comment:
what kind of options?

Reply:
the listing options. sorting, filtering etc.